File: Брославський Юрій/Customer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab_1
{
    internal class Customer
    {
        private int ID;
        private string Name;
        private int Age;
        private Operator[] Operators;
        private Bill Bill;

        public Customer(int id, string name, int age, Operator[] operators, Bill[] bills, double limitingAmount)
        {
            ID = id;
            Name = name;
            Age = age;
            Operators = operators;
            Bill = new Bill(limitingAmount);
            bills[id] = Bill;
        }

        public string GetName()
        {
            return Name;
        }

        public void Talk(int minute, Customer other)
        {
            Operator customerOperator = Operators[ID];
            Operator otherOperator = Operators[other.ID];

            double charge = customerOperator.CalculateTalkingCost(minute, this);

            if (Age < 18 || Age > 65)
            {
                charge -= charge * customerOperator.GetDiscountRate() / 100.0;
            }

            Bill.Add(charge);

            Console.WriteLine($"{Name} talked to {other.GetName()} for {minute} minutes. Charge: {charge:C}");
        }

        public void Message(int quantity, Customer other)
        {
            Operator customerOperator = Operators[ID];
            Operator otherOperator = Operators[other.ID];

            double charge = customerOperator.CalculateMessageCost(quantity, this, other);

            if (customerOperator == otherOperator)
            {
                charge -= charge * customerOperator.GetDiscountRate() / 100.0;
            }

            Bill.Add(charge);

            Console.WriteLine($"{Name} sent {quantity} messages to {other.GetName()}. Charge: {charge:C}");
        }

        public void Connection(double amount)
        {
            Operator customerOperator = Operators[ID];

            double charge = customerOperator.CalculateNetworkCost(amount);

            Bill.Add(charge);

            Console.WriteLine($"{Name} connected to the internet using {amount} MB. Charge: {charge:C}");
        }

        public void PayBill(double amount)
        {
            Bill.Pay(amount, GetName());
        }

        public void ChangeOperator(Operator newOperator)
        {
            Operators[ID] = newOperator;
            Console.WriteLine($"{Name} changed operator to Operator {newOperator.GetID()}");
        }

        public void ChangeBillLimit(double newLimit)
        {
            Bill.ChangeTheLimit(newLimit);
            Console.WriteLine($"{Name} changed the bill limit to {newLimit:C}");
        }

        public Bill GetBill()
        {
            return Bill;
        }
    }
}

File: Грабар Данило/lab1/customer.py
class Customer:
    def __init__(self, id, name, surname, age, bill, operator):
        self.id = id
        self.name = name
        self.surname = surname
        self._age = age
        self.bill = bill
        self._operator = operator

    def talk(self, minute, otherCustomer):
        self.bill._currDebt += self.operator.calculateTalkingCost(minute, currCustomer=self)
        return f"Customer {self.name} {self.surname} talked with customer {otherCustomer.name} {otherCustomer.surname} {minute} minutes."

    def message(self, quantity, otherCustomer):
        self.bill._currDebt += self.operator.calculateMessageCost(quantity, currCustomer=self, otherCustomer=otherCustomer)
        return f"Customer {self.name} {self.surname} sent {quantity} messages to {otherCustomer.name} {otherCustomer.surname}. "

    def connect(self, amount):
        if self.bill.check(amount):
            self.bill._currDebt += self.operator.calculateNetworkCost(amount)
            return f"Customer {self.name} {self.surname} connected to network ({amount} mbps)."
        else:
            return "Limit error"

    def changeOperator(self, id, operators):
        self.operator = operators[id]

    @property
    def age(self):
        return self._age

    @age.setter
    def age(self, newValue):
        self._age = newValue

    @property
    def operator(self):
        return self._operator

    @operator.setter
    def operator(self, newValue):
        self._operator = newValue
